### PR TITLE
fix(fusions): compose caller mask with sliding-window mask instead of dropping it

### LINKED ISSUE #7039: Sliding-window path in FusedScaleMaskSoftmax silently discards the caller-provided attention mask
## Summary

In `megatron/core/fusions/fused_softmax.py::forward_torch_softmax` the sliding-window branch **unconditionally overwrites** the caller-provided mask:

```python
# Generate causal mask if not given
sq, sk = input.size(2), input.size(3)
if self.window_size is not None:
    mask = get_sliding_window_causal_mask(sq, sk, self.window_size)   # <- discards the incoming mask
elif self.attn_mask_type == AttnMaskType.causal and mask is None and sq > 1:
    ...
    mask = get_default_causal_mask(sq)
```

Note the asymmetry: the plain-causal branch is correctly guarded by `mask is None`, the sliding-window branch is not.

## Impact

For any model with sliding-window layers, a padding mask (micro_batch_size > 1) or a packed-sequence block-diagonal mask passed by the caller is silently ignored on the SWA layers: pad tokens / cross-document tokens receive attention, the loss is quietly polluted, and nothing raises.

We hit this while running full SFT of MiMo-V2.5 (39 of 48 layers are SWA with this torch-fallback path): the only safe configurations today are micro_batch_size=1 and no sequence packing, which is a significant efficiency loss for variable-length multimodal data.

## Repro sketch

Build `FusedScaleMaskSoftmax(..., attn_mask_type=AttnMaskType.causal, scaled_masked_softmax_fusion=False, window_size=(4, 0))`, pass zero scores of shape [2, 1, 8, 8] and a bool padding mask masking the last 3 keys of sample 1 — the returned probabilities at the padded keys are non-zero.

## Proposed fix

Compose instead of overwrite (both masks are bool with True = masked; `[b,1,sq,sk] | [sq,sk]` broadcasts):

```python
swa_mask = get_sliding_window_causal_mask(sq, sk, self.window_size)
mask = swa_mask if mask is None else (mask | swa_mask)
```

Behavior is unchanged for every caller that passes `mask=None`. PR with this fix + a unit test incoming (scope: the torch fallback path only; fused/TE kernel paths are untouched).

### PR DESCRIPTION
**Pull request by yuweih205**
Fixes #7039

The PyTorch sliding-window softmax path silently discards caller-provided masks. Combine the caller mask with the generated sliding-window mask using `mask | sliding_window_mask`; both are boolean masks with `True` meaning masked, and broadcasting is preserved. With `mask=None`, use the generated window mask as before.

This restores the original composition behavior and regression test, with `|` in place of `torch.logical_or`. The temporary error for simultaneous `mask` and `window_size` has been removed. Scope remains the PyTorch fallback path.

Validation: the original focused regression passed in an isolated CPU harness with PyTorch 2.13.0, using the actual softmax and mask-builder source with CUDA allocations redirected to CPU. CUDA and fused kernels were not tested. Black, isort, pylint and Ruff passed; the advisory mypy step reports missing dependencies and the existing `SoftmaxOne` annotation issue.

### CODE REVIEW

File: megatron/core/fusions/fused_softmax.py [hunk 1]
```diff
@@ -319,7 +319,10 @@ def forward_torch_softmax(self, input, mask, softmax_offset=None):
         # Generate causal mask if not given
         sq, sk = input.size(2), input.size(3)
         if self.window_size is not None:
-            mask = get_sliding_window_causal_mask(sq, sk, self.window_size)
+            swa_mask = get_sliding_window_causal_mask(sq, sk, self.window_size)
```
Comment (wujingyue) [inline megatron/core/fusions/fused_softmax.py]: Don't abbreviate here -- bad readability.

File: megatron/core/fusions/fused_softmax.py [hunk 2]
```diff
@@ -319,7 +319,10 @@ def forward_torch_softmax(self, input, mask, softmax_offset=None):
...
+            # Compose with a caller-provided (e.g. padding or packed-sequence) mask
+            # instead of silently discarding it; both are bool with True = masked.
+            mask = swa_mask if mask is None else (mask | swa_mask)
```
Comment (wujingyue) [inline megatron/core/fusions/fused_softmax.py]: `|` feels like a surprising contract. However, I'll defer to the owner.

File: megatron/core/fusions/fused_softmax.py (hunk 1 above)
Comment (yuweih205) [reply megatron/core/fusions/fused_softmax.py]: Thanks a lot for pointing this out, and sorry about the unclear abbreviation. I have renamed swa_mask to sliding_window_mask in 306b306cf so the intent is easier to follow.

File: megatron/core/fusions/fused_softmax.py (hunk 2 above)
Comment (yuweih205) [reply megatron/core/fusions/fused_softmax.py]: Thanks a lot for flagging this. You are right that the bare bitwise-OR expression makes the mask contract harder to read than it should be. I changed it to torch.logical_or(mask, sliding_window_mask) in 306b306cf. Both tensors here are boolean masks, with True meaning masked, and broadcasting stays the same. If the owner would prefer a different way to compose them, I would be very happy to adjust it.

Comment (wujingyue) [reply megatron/core/fusions/fused_softmax.py]: I was probably unclear. 

I wasn't judging `|` vs `logical_or` -- in fact, `|` looks better just between the two. 

I was considering alternatives like throwing a hard error when both `mask` and sliding window are enabled. `mask` seems to say that the user wants to completely override the mask calculation.

Comment (yuweih205) [reply megatron/core/fusions/fused_softmax.py]: OK, switched back to `|`.

Comment (yuweih205) [reply megatron/core/fusions/fused_softmax.py]: If the user-provided mask is meant to completely override the mask calculation, shouldn't we use it directly and skip generating the SWA mask, rather than raise an error when both are configured?

**Review by wujingyue — APPROVED** [review]

**Comment by wujingyue** [comment]
Thanks for waiting. I pinged @NVIDIA/core-adlr and @NVIDIA/core-nemo again internally

**Review by ericharper — APPROVED** [review]

**Comment by yuweih205** [comment]
The red `linting` job was the new kernel-determinism coverage check, not a lint error: `fused_softmax.py` is registered as the `fused_scale_mask_softmax` kernel in `tests/unit_tests/determinism/kernels/manifest.py`, so a change to it must come with a change to its determinism test.

Pushed `59c5b74bf`: `test_sliding_window_softmax_torch_path_replays` in `test_tensor_parallel_kernels.py` — a bit-exact replay of the `window_size` torch path (the branch this PR touches), parametrized over the sliding-window mask alone and composed with a caller padding mask. `tools/check_kernel_determinism_coverage.py --base-ref main` now passes locally on this change set; the replay itself was checked on an H200 (both cases bit-exact across replays, and padded keys get ~0 probability on rows that keep a live key). The other failing checks are the approval gates downstream of the lint job.

@wujingyue could you `/ok to test 59c5b74bfe32ed614f0058da963710dd4e22e13e`? The linting job should go green with the determinism test in place.

**Comment by yuweih205** [comment]
@deepakn94 Could you (or another `@NVIDIA/core-adlr` codeowner) take a look when you have a moment? The `codeowners-approval` check reports the only missing piece is a core-adlr approval for `megatron/core/fusions/fused_softmax.py`; @wujingyue and @ericharper have approved, all tests on `59c5b74b` are green, and the change is a small one (compose the caller mask with the sliding-window mask instead of dropping it, plus the determinism replay test). Thanks!

**Comment by yuweih205** [comment]
@NVIDIA/core-adlr Could a code owner please review this PR for approval?

The change composes the caller mask with the sliding-window mask in `megatron/core/fusions/fused_softmax.py`, with regression and determinism coverage. On the current head (`59c5b74b`), lint, unit tests, and the executed integration tests are green. wujingyue and ericharper have already approved; the remaining `codeowners-approval` failure specifically requests an approval from `@NVIDIA/core-adlr`.

The team is already listed under requested reviewers. I would appreciate a review when someone is available, and am happy to address any remaining concerns. Thanks!

**Review by cuichenx — APPROVED** [review]
LGTM. However the TE path is much more performance. Is the TE path missing a functionality for you?

**Review by jaredcasper — APPROVED** [review]

**Comment by yuweih205** [comment]
@cuichenx Thanks! TE is not missing SWA/padding support here, and I agree it is the preferred path for performance.

This PR is mainly a correctness fix for the native PyTorch fallback. `DotProductAttention` still exposes this path, and with `window_size` enabled it currently drops a caller-provided mask. We hit this in MiMo-V2.5 SFT with variable-length batches, where padding/packed-sequence masking becomes incorrect.

So the goal here is not to make the PyTorch path competitive with TE, but simply to keep the existing fallback semantically correct.

**Comment by yuweih205** [comment]
@wujingyue Could you please re-add this PR to the merge queue?

The previous merge-group run was removed after the SSM unit-test job exceeded its 1-hour limit. The pytest session itself completed with **511 passed, 9 skipped** in 56:16; the timeout happened during the remaining post-test work. The same 520-test SSM bucket passed in the surrounding merge-group runs at roughly 55–56 minutes, and the SSM coverage did not exercise this PR's `fused_softmax.py` change, so this appears to be runner/runtime variance rather than a test failure.

Failed run: https://github.com/NVIDIA/Megatron-LM/actions/runs/35680878378/job/106613879245

Thank you!
